Consolidate RDF and RDG tests (part 1)

captainsafia opened the pull request:
Part of https://github.com/dotnet/aspnetcore/issues/46196.

This PR also resolves a bug in the RDG. When the user annotates a parameter as `FromBody` but does not actually provide a property (`Content-Length` is 0 and `CanHaveBody` is set to `false`), the `TryResolveBody` method in the RDG was not setting the status code to 400 to match the behavior of RDF.

captainsafia committed:
Consolidate RDF and RDG tests (part 1)

adityamandaleeka commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/Baselines/MapAction_ExplicitSource_SimpleReturn_Snapshot.generated.txt:
```diff
@@ -534,6 +534,12 @@ namespace Microsoft.AspNetCore.Http.Generated
                 }
             }
+
+            if (!allowEmpty)
```
First time looking at this code/area so maybe I missed something, but this is logically an `else if` right? Doesn't seem to be anyway to fall out of the prior `if` block once you're in it.

If so, consider making this an `else if` for clarity (here and in other places).

captainsafia replied:
Yep, this is the designed to handle the case where you've specified that a JSON body is required in this request (`allowEmpty = false`) but the `IHttpRequestBodyDetectionFeature` didn't detect a body in the request because there was no content length set.

adityamandaleeka reviewed (approved):
LGTM other than the minor comment.

JamesNK commented:
@captainsafia I'm guessing a lot of RDF tests currently aren't supported by RDG. What are you doing there? Are you not moving those tests yet, or do you have a mechanism to add them now but disable them just for RDG?

I believe with the routing tests, if a tested feature was only supported by one of the routing implementations, then we'd make the test virtual, and then override it in one of the implementations and add `[Fact]` (or it was the other way around, and we added skip, it was so long ago I don't remember)

captainsafia commented:
> do you have a mechanism to add them now but disable them just for RDG?

With the way things are set up now:

- Tests that are unique to the RDF are written in [RuntimeGeneratorTests](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/test/RequestDelegateGenerator/RuntimeGeneratorTests.cs)
- Tests that are unique to the RDG are written in [CompileTimeGeneratorTests](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/test/RequestDelegateGenerator/CompileTimeGeneratorTests.cs)
- Tests that are shared between the two are implement in the partial abstract base class ([ref](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTestBase.cs))

> I'm guessing a lot of RDF tests currently aren't supported by RDG. What are you doing there?

At the moment, I am prioritizing tests that can be shared between the RDF and RDG so that we can catch bugs in the RDG for features that have already been implemented.

The second order of priority is tests that are unique to RDF which will be moved to the `RuntimeGeneratorTests`. implementation. As we add more features to the RDG, they will be moved from the `RuntimeGeneratorTests` class to the shared base class. We'll probably have some tests that we don't move due to limitations in the source generator implementation.

There's a third class of tests that will stay in the `RequestDelegateFactoryTest` class because they aren't behavior tests that validate the behavior of the generated `RequestDelegate` but those aren't a majority of the tests.

captainsafia commented:
@mitchdenny Fancy a review on this? I moved some tests over and also split up the base class into multiple files so we avoid having a long test file.

This also includes your latest implicit TryParse changes.

JamesNK commented on src/Http/Http.Extensions/test/RequestDelegateGenerator/RequestDelegateGeneratorTests.JsonBody.cs:
```diff
@@ -0,0 +1,151 @@
+using Microsoft.AspNetCore.Http.Features;
+namespace Microsoft.AspNetCore.Http.Generators.Tests;
+
+public abstract partial class RequestDelegateGeneratorTests
```
I think renaming this class would reduce confusion. Perhaps `RequestDelegateCreationTests`?

(not urgent: can be in its own PR in the future)

captainsafia replied:
Naming is hard! Thanks for the recommendation. I'll admit "Strategy" was probably too posh 😛  I went ahead and renamed it.

mitchdenny commented:
> @mitchdenny Fancy a review on this? I moved some tests over and also split up the base class into multiple files so we avoid having a long test file.
> 
> This also includes your latest implicit TryParse changes.

Yeah, looks good. I looked earlier today but got distracted. When this goes in I think my PR will have a merge conflict but it should mostly be a matter of cutting and pasting some tests around. Keen to get this in so we can start to lean on more of the existing test cases to verify what we've done.

captainsafia committed:
Merge remote-tracking branch 'origin/main' into safia/rdg-tests-migrate

captainsafia committed:
Fix else-if in TryResolveBodyAsync

captainsafia committed:
Rename to RequestDelegateCreation tests

captainsafia committed:
Undo accidental base file changes
